libraries/ms-common/src/stm32f0xx/soft_timer.c

```diff
@@ -0,0 +1,211 @@
+static volatile uint32_t s_active_timer_duration;
+
+static void prv_start_timer(uint32_t duration_us) {
+  // Clear any pending interrupts, this will always occur in an interrupt or new timer getting
+  // started in a critical section so we need to prevent a accidental trigger if the time expires in
+  // the middle of the critical section and we started a new timer which falls before it.
+  TIM_ClearITPendingBit(TIM2, TIM_IT_CC1);
+
+  TIM_Cmd(TIM2, DISABLE);
+  uint32_t curr_time = TIM_GetCounter(TIM2);
+  if (curr_time + duration_us > UINT32_MAX) {
+    TIM_SetCompare1(TIM2, curr_time + duration_us - UINT32_MAX);
+  } else {
+    TIM_SetCompare1(TIM2, curr_time + duration_us);
+  }
+  s_active_timer_duration = duration_us;
+  TIM_Cmd(TIM2, ENABLE);
+}
+
+static uint32_t prv_soft_timer_update_period(void) {
+  uint32_t counter = TIM_GetCounter(TIM2);
+  if (counter - s_last_update_count > 0) {
+    return counter - s_last_update_count;
+  } else {
+    return counter - s_last_update_count + UINT32_MAX;
+  }
+}
+
+static void prv_soft_timer_update(uint32_t time_since_update) {
+  // Default these values to max. If there is ever a newer timer it will replace them. In the event
+  // there are no timers left the defaults will result in the module being reset to a state where no
+  // timers exist.
+  s_last_update_count = TIM_GetCounter(TIM2);
+  SoftTimerID min_time_id = SOFT_TIMER_MAX_TIMERS;
+  uint32_t min_duration = UINT32_MAX;
+
+  for (uint32_t i = 0; i < SOFT_TIMER_MAX_TIMERS; i++) {
+    // If the timer is in use and wasn't just added then update its duration.
+    if (s_soft_timer_array[i].inuse) {
+      if (time_since_update >= s_soft_timer_array[i].duration) {
+        // For each inuse timer see if the timer expired. If it has run its callback.
+        s_soft_timer_array[i].callback(i, s_soft_timer_array[i].context);
+        s_soft_timer_array[i].inuse = false;
+      } else {
+        // Otherwise, update the duration left on the timer.
+        s_soft_timer_array[i].duration -= time_since_update;
+        // Figure out the next timer.
+        if (min_duration >= s_soft_timer_array[i].duration) {
+          min_time_id = i;
+          min_duration = s_soft_timer_array[i].duration;
+        }
+      }
+    }
+  }
+  // Set the active timer id.
+  s_active_timer_id = min_time_id;
+}
+
+void soft_timer_init(void) {
+  // Start the PeiphClock
+  RCC_APB1PeriphClockCmd(RCC_APB1Periph_TIM2, ENABLE);
+  // Stop the timer if it was running and ensure only overflows trigger updates.
+  // Also remove the ARR preload.
+  TIM_Cmd(TIM2, DISABLE);
+  // TIM_InternalClockConfig(TIM2);
+
+  // Get the clock speed of the clocks to determine the SYSCLK speed which is what TIM2 uses.
+  RCC_ClocksTypeDef clock_speeds;
+  RCC_GetClocksFreq(&clock_speeds);
+
+  // Configure each clock tick to be 1 microsecond from 0 to duration.
+  TIM_TimeBaseInitTypeDef init_struct = { .TIM_Prescaler = clock_speeds.PCLK_Frequency / 1000000,
```

Comment:
From the stm32f0xx library: stm32f0xx_rcc.c line 19:

> After reset the device is running from HSI (8 MHz) with Flash 0 WS, 
         all peripherals are off except internal SRAM, Flash and SWD.
         (#) There is no prescaler on High speed (AHB) and Low speed (APB) busses;
             all peripherals mapped on these busses are running at HSI speed.

TIM2 runs on APB bus 1 so this should be valid.

This implies I can just use SYSCLK and divide by 1000000 to get the prescaler to use. This seems consistent with running the timer for a longer period say 10s and comparing to a stopwatch. I am aware these values can be changed and we will probably change them but until we make that change this is technically correct. Unless you modified the defaults somewhere in the linker or something? In which case please document :p